dds_web/database/models.py

```diff
@@ -68,6 +68,38 @@ class ProjectUserKeys(db.Model):
     key = db.Column(db.LargeBinary(300), nullable=False, unique=True)
 
 
+class ProjectInviteKeys(db.Model):
+    """
+    Many-to-many association table between projects and invites.
+
+    Primary key(s):
+    - project_id
+    - invite_id
+
+    Foreign key(s):
+    - project_id
+    - invite_id
+    """
+
+    # Table setup
+    __tablename__ = "projectinvitekeys"
+
+    # Foreign keys & relationships
+    project_id = db.Column(
+        db.Integer, db.ForeignKey("projects.id", ondelete="CASCADE"), primary_key=True
+    )
+    project = db.relationship("Project", back_populates="project_invite_keys")
+    # ---
+    invite_id = db.Column(
+        db.Integer, db.ForeignKey("invites.id", ondelete="CASCADE"), primary_key=True
+    )
+    invite = db.relationship("Invite", back_populates="project_invite_keys")
+    # ---
+
+    # Additional columns
+    key = db.Column(db.LargeBinary(300), nullable=False, unique=True)
+
```

Comment:
I think the `ProjectInviteKeys` need an additional owner column analogous to the `ProjectUser] table. I would actually suggest to duplicate this information also in the `ProjectUserKeys`, because then we could easily check if a Researcher is allowed to share (Project Owner) the project key or not.

Reply:
Thank you for bringing up an important detail!
I double checked a few places. Here are some things to raise:

- According to the role responsibilities agreement that we have established, a research account cannot invite someone to a project. The minimum required level of role for inviting to a project is project owner.
- The implementation to invite to a project does not take the previous point into account. I wrote a new test case for it and it fails, meaning that it is now possible that a researcher can invite someone to a project. I think we need to create a bug for this. I can immediately fix it in that, if agreed. @inaod568 @MatthiasZepper @alneberg 
- Considering that we can fix this in the invite function, we do not need to keep track of who has invited who under normal operation. However I highly recommend that we log who has invited who to which project at which date and time as part of our audit log. This can be investigated from the log if needed.
- So all in all, I don't think we need to add a new column to the model.

Reply:
>     * According to the role responsibilities agreement that we have established, a research account cannot invite someone to a project. The minimum required level of role for inviting to a project is project owner.

That is correct. A project owner status is required to grant further users access to a project.

I was a bit sloppy in my previous comment, because I just wrote _Researcher_ (the return value of `User.role`), even though I meant `ResearchUser` with regard to the database models.

The logical problem is, that the role of a Unit Admin is a property of the user, such that we can return either a _Unit Personnel_ or _Unit Admin_ value for `User.role` on `UnitUsers`(database model), but for the researchers it is more complicated as they can be the owners of just some projects they have access to.

>     * The implementation to invite to a project does not take the previous point into account. I wrote a new test case for it and it fails, meaning that it is now possible that a researcher can invite someone to a project. I think we need to create a bug for this. I can immediately fix it in that, if agreed. @inaod568 @MatthiasZepper @alneberg

If you wish to fix that, I am happy to leave it to you. 

In a way I was aware of this and wanted to sort that out, but by means of by adding that owner column to the keys. I have no real understanding how to straightforwardly solve it otherwise.  

>     * Considering that we can fix this in the invite function, we do not need to keep track of who has invited who under normal operation. However I highly recommend that we log who has invited who to which project at which date and time as part of our audit log. This can be investigated from the log if needed.

Agree. With the structlog package in place, this should be fairly straightforward. See example for the deletion here:

https://github.com/ScilifelabDataCentre/dds_web/blob/cfa6bfc001205ee645dff3cc1361edc99f17cafc/dds_web/api/user.py#L519-L523

>     * So all in all, I don't think we need to add a new column to the model.

Not convinced yet, but I also don't insist on it.

Reply:
As far as I can follow everyone's point:
`InviteUserSchema` in `invite_user` loads the **role** argument when someone tries to invite someone. This is persisted into the db as `invite.role`, so if a project owner is invited `invite.role = "Project Owner"`. Then we generate key pair for invite and invite.id is associated with the generated keys in `ProjectInviteKeys`. Invited user goes to confirm_invite with the token, we make sure that the invite is in the db and fetch it. At this point you have the invite with its role already set and there is an association to the generated keys for itself, so as I see it the role is immediately accessible.
Please help me understand if I'm missing someone's point. We could perhaps do a spontaneous chat about it, if needed?

Reply:
Aahh! I follow your meaning now. Yes, this should be already handled then! @alneberg @MatthiasZepper

Reply:
We want to be able to invite a user to multiple projects at once where the user is an owner of some but not an owner of others. I think we have called "Project Owner" a user role a bit too sloppily. It's not a user role, it's a user-project specific permission. But anyway. Inviting to a project is not even implemented yet so this can all wait until later. I think it's important to get these changes merged and we can build from here instead of creating huge PRs.

Reply:
Yes we talked about inviting to multiple projects this week. It is a separate issue to solve. And yes again that right now we don't pass in a project arg to the invite so it is not helpful to set role to project owner while inviting. I think we need to create specific issues for these and write clear a description for each.

Reply:
https://github.com/scilifelabdatacentre/dds_web/issues/860

Reply:
Feel free to add more info if you want to.

Reply:
I fully agree with being everything said. 

The case that _an invited e-mail needs to have owner privileges for some, but not all projects_  was indeed the biggest issue I had during my work. Even worse, Phil had specifically requested that one needs to be able to specify multiple projects in one invocation of the `dds user add` command in our BioInfo meeting on Monday.

Since #798 was basically a "Fix everything" issue, I felt responsible for all of it. Solving the problem of _Fine-grained project ownership to an invitee_ was basically what I felt would be the purpose of #825?

Anyway: I am more than happy to see this PR merged and then build on top of it whatever still remains to be done! Because the problem of being confronted with an overarching task was exactly what caused me a lot of headache in the past days.

Reply:
@MatthiasZepper I agree that the full task of inviting a user to a project, also being able to invite a user to multiple projects (has been on the table for a while), also adding existing users to projects and having different procedures depending on which role the invited user is and the current user has, and to add the project keys to the invite -- that's a big task. However, the task that you were assigned to/agreed to do (#798) has a description that I don't see as a "fix everything" and I also have tried to explain what I mean during at least two of our meetings last week. In this case it turned out that it was better to do certain other tasks before approaching the task you started with, which sometimes is what happens. If you feel that you have a tough task ahead of you and are unsure about what the goal is, please bring this up and discuss this as soon as possible so that we can sort out any confusion (maybe theres a misinterpretation on my part, or yours) and possibly even change which task you work on. 👍🏻

Reply:
> Just a side note, @MatthiasZepper a single invocation of `dds user add` can still entitle several requests to the API so it doesn't have to mean that the endpoint needs to be able to accept multiple projects at once.

True. That is actually the much easier way to go and solve that in the CLI. I can do that.